### ISSUE TITLE
Incorrect ground truths

### ISSUE THREAD

**Issue by andreas-abel** [ithemal/Ithemal#54]
In table 1 in the paper, you claim that the actual execution time (which you consider to be the ground truth) for 100 iterations of the code sequence `shl rbx, 0x02; mov rdi, rbx` is 83 core clock cycles. This is impossible. The `shl` instruction both reads and writes the `rbx` register. Thus, there is a read after write dependency, which means that the execution time of 100 iterations cannot be less than 100 cycles.

If I benchmark the sequence with nanoBench (https://github.com/andreas-abel/nanoBench), I get exactly 100 cycles, which is also close to the value IACA reports. Also, if I benchmark the first sequence in the table (`vxorps xmm0, xmm0, xmm0`) with nanoBench, I get 25 cycles, which is also much closer to IACA's than to your results.

There are several problems with your timing measurement code that lead to these incorrect results. 

First of all, you should note that the `rdpmc` instruction is not a serializing instruction. Thus, you need to add serializing instructions before and after executions of the `rdpmc` instruction. The code in `timing_tools/harness/run_test.nasm` lacks such a serializing instruction between the `rdpmc` call in line 261 and `test_impl` in line 271. 

Another problem with your code is that (in `gettiming.py`) you determine the execution time by measuring the time of 100 iterations of the sequence and then subtract the time of an empty sequence. This does not lead to meaningful results, as part of the execution time of the 100 iterations gets hidden by the overhead of the `cpuid` instruction. The `cpuid` is a serializing instruction, but note that this does not guarantee that all previous instructions must have finished by the time the `cpuid` instruction begins. It only guarantees that all previous instructions have finished by the time the next instruction after the `cpuid` instruction begins execution. If you instead measure the time for, e.g., 200 iterations, and then subtract the time for 100 iterations, you will see that difference will be close to 100 cycles, which matches the nanoBench and IACA results.

Since your ground truths are incorrect, this makes your claims later in the paper that Ithemal is more accurate than IACA questionable. Can you repeat your experiments with a more accurate way of obtaining ground truths, such as nanoBench, or by fixing your measurement infrastructure?

**Comment by CharithYMendis** [comment]
The timing_harness is outdated. So your comparisons are invalid. We have another paper coming up with the timing harness we used. Please be patient till we release the timing tools.

**Comment by andreas-abel** [comment]
But the results in table 1 in the paper are incorrect, independent of which timing harness was used.

**Comment by ychen306** [comment]
Hi Andreas,
Thanks for pointing out the error. We have fixed test harness and updated the formula (we actually do use the 200 - 100 formula) earlier. The 83 number is a result of a bug that was then fixed. The timing harness is out-of-date, and we don't use it anymore. We retrained ithemal with the not-yet-released timing number, and it holds. We are releasing the data set and our timing tool in the near future, together with a paper detailing our methodology in IISWC.
